src/main/java/net/snowflake/client/core/SFSession.java

```diff
@@ -586,7 +606,7 @@ protected void startHeartbeatForThisSession() {
       logger.debug("start heartbeat, master token validity: " + masterTokenValidityInSeconds);
 
       HeartbeatBackground.getInstance()
-          .addSession(this, masterTokenValidityInSeconds, this.heartbeatFrequency);
+          .addSession(this, masterTokenValidityInSeconds, heartbeatFrequency);
```

Comment:
So it looks like if one session (connection) has set the heartbeatFrequency, then all sessions will use this heartbeatFrequency. I guess this is what we want?

Reply:
This addSession() function adds 1 session at a time. The heartbeat frequency is set inside the session. So the heartbeat frequency will just be changed for this particular session, which is what we want. Once a new session is started, the defaults will be used.

Reply:
Hi Meg, I just double checked HeartbeatBackground.java again and it looks like HeartbeatBackground is a Singleton. All sessions added in it share the same heartbeatFrequency, i.e. all sessions emit heat beat at "the same time" (in one for loop). So it appears to me that this heartbeatFrequency is like a "global" settings even though it can be set for each session itself?

Reply:
Ah, I guess the heartbeat frequency would have to be the same for if you had a connection pool with multiple sessions/connections for a single driver. So it's sort of "global", but only for a given JVM, I believe. This logic was already there and I don't want to change this part of it.

Reply:
Yeah, I guess this would not hurt us a lot. I don't think we need to change the logic as well, at least for this time. Just wanted to double check we know the exact behavior